src/Microsoft.DocAsCode.Dfm/DfmCodeExtractor.cs

```diff
@@ -233,42 +233,81 @@ public DfmExtractCodeResult ExtractFencesCode(DfmFencesBlockToken token, string
+                return GetFencesCodeCore(fencesCode, token.PathQueryOption.Regions, maxCommonIndent);
+            }
+            else
+            {
+                // Add the full file when no query option is given
+                int indentLength = (from line in fencesCode
                                     where !string.IsNullOrEmpty(line) && !string.IsNullOrWhiteSpace(line)
                                     select (int?)DfmCodeExtractorHelper.GetIndentLength(line)).Min() ?? 0;
-                return GetFencesCodeCore(fencesCode, startLine, endLine, indentLength);
+                return GetFencesCodeCore(fencesCode, 1, fencesCode.Length, indentLength);
             }
         }
 
+        private DfmExtractCodeResult GetFencesCodeCore(string[] codeLines, List<DfmFencesBlockPathQueryRegion> regions, int indentLength)
+        {
+            var includedLines = new List<string>();
+            foreach (var region in regions)
+            {
+                int startLine = region.StartLine ?? 1;
+                int endLine = region.EndLine ?? codeLines.Length;
+                AddFencesCodeRegion(includedLines, codeLines, startLine, endLine, indentLength);
+            }
+
+            return new DfmExtractCodeResult
+            {
+                IsSuccessful = true,
+                FencesCodeLines = TrimIndent(includedLines, indentLength)
+            };
+        }
+
         private DfmExtractCodeResult GetFencesCodeCore(string[] codeLines, int startLine, int endLine, int indentLength, HashSet<int> excludedLines = null)
         {
-            long totalLines = codeLines.Length;
             var includedLines = new List<string>();
-            for (int i = startLine; i <= Math.Min(endLine, totalLines); i++)
+            AddFencesCodeRegion(includedLines, codeLines, startLine, endLine, indentLength, excludedLines);
+
+            return new DfmExtractCodeResult
+            {
+                IsSuccessful = true,
+                FencesCodeLines = TrimIndent(includedLines, indentLength),
+            };
+        }
+
+        private void AddFencesCodeRegion(List<string> includedLines, string[] codeLines, int startLine, int endLine, int indentLength, HashSet<int> excludedLines = null)
+        {
+            for (int i = startLine; i <= Math.Min(endLine, codeLines.Length); i++)
             {
                 if (excludedLines == null || !excludedLines.Contains(i))
                 {
                     includedLines.Add(codeLines[i - 1]);
                 }
             }
+        }
 
-            return new DfmExtractCodeResult
-            {
-                IsSuccessful = true,
-                FencesCodeLines = (indentLength == 0 ? includedLines : includedLines.Select(s => Regex.Replace(s, string.Format(RemoveIndentSpacesRegexString, indentLength), string.Empty))).ToArray()
-            };
+        private string[] TrimIndent(List<string> codeLines, int indentLength)
```

Comment:
what is the indent trim principle? *Collect all the ranges and then calculate the indent together* or *calculate indent for each region and than collect the trimmed ranges*?